### PR TITLE
keep integration (4 widgets) working with Drupal 8 version (under development) of OpenScholar #6381

### LINKED ISSUE #6381: keep integration (4 widgets) working with Drupal 8 version (under development) of OpenScholar
Hey guys, we are updating the dataverse widgets for the D8 version of our openscholar code and were hoping you could make some small changes to (https://dataverse.harvard.edu/resources/js/widgets.js).

See: https://github.com/openscholar/openscholar/issues/12452#issuecomment-555427431

We are passing the params in URL encoded now and were hoping you could add decoding to your JS code so that everything works as expected.  Please let me know if you have any questions.

@stamina84

### PR DESCRIPTION
**Pull request by pdurbin**
**What this PR does / why we need it**:

OpenScholar is upgrading to Drupal 8 and some of the Dataverse widgets aren't working. Maybe all.

**Which issue(s) this PR closes**:

Closes #6381

**Special notes for your reviewer**:

This is the tiniest fix I could make to fix a simple non-OpenScholar test I tried with these two variations:

## original

`<script src="http://localhost:8080/resources/js/widgets.js?persistentId=doi:10.5072/FK2/6YRLFR&amp;dvUrl=http://localhost:8080&amp;widget=iframe&amp;heightPx=500"</script>`

## encoded colons

`<script src="http://localhost:8080/resources/js/widgets.js?persistentId=doi%3A10.5072/FK2/6YRLFR&amp;dvUrl=http%3A//localhost:8080&amp;widget=iframe&amp;heightPx=500"</script>`

**Suggestions on how to test this**:

- Deploy this branch to a server
- Contact OpenScholar and ask them to point a test server at your Dataverse deployment

**Does this PR introduce a user interface change?**:

No. In fact, I didn't touch the Dataverse UI where people can copy with widgets from. The old URL with a colon (:) still seemed to work  from my simply test.

**Is there a release notes update needed for this change?**:

Maybe something like "If you are running OpenScholar future version whatever, Dataverse widgets do not work. You must be running this version of Dataverse or newer."

**Additional documentation**:

With no changes to Dataverse, I was getting this error with the "encoded colons" version above:

Screen Shot 2020-01-06 at 3 59 45 PM

With this pull request, it seems to work, at least for this widget:

Screen Shot 2020-01-06 at 4 00 00 PM

### CODE REVIEW

**Comment by stamina84** [comment]
@pdurbin I see you replace only colon char, what if you use this function to decode the whole string? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/decodeURIComponent

**Comment by pdurbin** [comment]
@stamina84 thanks, I did think about that and could try it. At the moment the problem is that I don't have a way to test any changes I'm making with OpenScholar. This is because OpenScholar is hard coded to https://dataverse.harvard.edu as we've been discussing at https://github.com/IQSS/dataverse/issues/6381#issuecomment-571088325

**Comment by stamina84** [comment]
Ok, thanks! Then lets Bilsi discuss it with Richard and see whats the next step.

**Comment by rbran100** [comment]
@pdurbin we are going to add a UI component to allow for this to be changed in a vsite, didn't realize there were so many DV installs live since Harvard mostly uses just the one.

**Comment by rbran100** [comment]
@pdurbin any way to validate the URL is actually a dataverse install?

**Comment by pdurbin** [comment]
> any way to validate the URL is actually a dataverse install?

@rbran100 I usually tell people to got to the "version" API endpoint like https://demo.dataverse.org/api/info/version to check for what I guess I'd call a Dataverse "fingerprint". The output should look something like this (but not pretty printed):

```
{
  "status": "OK",
  "data": {
    "version": "4.18.1",
    "build": "267-a91d370"
  }
}
```

**Comment by djbrooke** [comment]
Thanks @rbran100 @stamina84, we'll remove this PR from our project board for now and will revisit once the new UI is in place in OS.

**Comment by pdurbin** [comment]
I deployed this pull request my "dev2" server and tried all four widgets. I'm numbering them like this: 

1. Dataset Citation
2. Dataset
3. Dataverse Search Box
4. Dataverse Listing

To make this a little more real, here's a screenshot of all four widgets on a vanilla HTML page:

chat dataverse org_error-documents_widgets html_-_2020-01-10_15 36 45

And here's the vanilla HTML I used:

```
<h2>1. Dataset Citation</h2>

<script src="https://dev2.dataverse.org/resources/js/widgets.js?persistentId=doi:10.5072/FK2/ASZCUR&amp;dvUrl=https://dev2.dataverse.org&amp;widget=citation&amp;heightPx=350"></script>

<hr>

<h2>2. Dataset</h2> 

<script src="https://dev2.dataverse.org/resources/js/widgets.js?persistentId=doi:10.5072/FK2/ASZCUR&amp;dvUrl=https://dev2.dataverse.org&amp;widget=iframe&amp;heightPx=800"></script>

<hr>

<h2>3. Dataverse Search Box</h2> 

<script src="https://dev2.dataverse.org/resources/js/widgets.js?alias=dataverse-project&amp;dvUrl=https://dev2.dataverse.org&amp;widget=search&amp;text=Search+my+dataverse"></script>

<hr>

<h2>4. Dataverse Listing</h2> 

<script src="https://dev2.dataverse.org/resources/js/widgets.js?alias=dataverse-project&amp;dvUrl=https://dev2.dataverse.org&amp;widgetScope=dataverse-project&amp;widget=iframe&amp;heightPx=800"></script>
```

Now, on to OpenScholar.

My observation is that the first two widgets don't work in OpenScholar without the fix in this pull request:

1. Dataset Citation
2. Dataset

That said, I just got off a Zoom with @Bilsi where we talked about some strange behavior within OpenScholar. Namely:

- While I can remove "Custom Text/HTML" block by dragging it, it is not possible to remove any of the Dataverse widgets by dragging or any other means. I demoed this to @Bilsi and she will open a bug on the OpenScholar side for this.
- In one case I created a page with the same title as a page we had just deleted and instead of the page being blank, it has a widget in it that I had been testing with the page I had deleted.

In short, we are very aware of bugs in layout and such in the Drupal 8 version of OpenScholar that's still under development. These bugs should be expected when testing this pull request.

**Comment by stamina84** [comment]
@pdurbin Thanks for the testing, I did a tests too, and seems good for me. The query url is "decoded" and embed properly. I think it can be merge if you agree

**Comment by stamina84** [comment]
@pdurbin could you please inform me, if its availiable at https://dataverse.harvard.edu/dataverse/ install? This is what we are using primary, thanks!

**Comment by pdurbin** [comment]
@stamina84 no, sorry, if you look at the bottom right corner of https://dataverse.harvard.edu (screenshot below), you'll see that it's still running 4.18.1.

Screen Shot 2020-01-16 at 7 20 42 AM

In #6506 we just started writing release notes for 4.19 and you are welcome to take a look and give feedback. This pull request will be part of 4.19.

Our release and deploy process is something like this:

- make a release https://github.com/IQSS/dataverse/releases
- time passes, maybe a week, it varies
- deploy new release to https://demo.dataverse.org
- deploy new release to https://dataverse.harvard.edu

I hope this helps! Also, thanks for working on the following issue that I opened!

- confusion over what to enter for "Dataset Persistent Identifier" in "Add new 'Dataverse Dataset' Widget" https://github.com/openscholar/openscholar/issues/12822

I know @jggautier @TaniaSchlatter and I have talked a lot about a related issue within Dataverse itself (users not knowing if they should enter "doi:" or not) but I can't find it right now.

**Comment by stamina84** [comment]
Ok, thanks for you information, its clear now!
